packages/lesswrong/components/ea-forum/users/EAGApplicationImportForm.tsx

```diff
@@ -0,0 +1,702 @@
+  const howOthersCanHelpMeRef = useRef<EditorFormComponentRefType>(null)
+  const howICanHelpOthersRef = useRef<EditorFormComponentRefType>(null)
+  // CKEditor fields use this to insert their data before we submit the form
+  const submitFormCallbacks = useRef<Array<Function>>([])
+  // CKEditor fields use this to clear local storage after successfully submitting the form
+  const successFormCallbacks = useRef<Array<Function>>([])
+  
+  const updateCurrentUser = useUpdateCurrentUser()
+  
+  const importMapLocation = async () => {
+    if (mapsLoaded && importedData?.mapLocation) {
+      // get a list of matching Google locations for the current lat/lng (reverse geocoding)
+      const geocoder = new googleMaps.Geocoder()
+      const geocodingResponse = await geocoder.geocode({
+        address: importedData.mapLocation
+      })
+      const results = geocodingResponse?.results
+      if (results?.length) {
+        return pickBestReverseGeocodingResult(results)
+      }
+    }
+    return null
+  }
+  
+  const importLinkedinProfileURL = () => {
+    // try to get the relevant snippet of the linkedin profile URL
+    const matches = importedData?.linkedinProfileURL?.match(/linkedin\.com\/in\/(\S+)/i)
+    return matches?.length ? matches[1] : ''
+  }
+  
+  const handleCopyAll = async (e) => {
+    e.preventDefault()
+    if (!importedData) return
+
+    // @ts-ignore TODO
+    setFormValues({
+      ...importedData,
+      mapLocation: await importMapLocation(),
+      linkedinProfileURL: importLinkedinProfileURL()
+    })
+    // update CKEditor fields
+    biographyRef?.current?.setEditorValue(importedData.biography.markdownValue)
+    howOthersCanHelpMeRef?.current?.setEditorValue(importedData.howOthersCanHelpMe.markdownValue)
+    howICanHelpOthersRef?.current?.setEditorValue(importedData.howICanHelpOthers.markdownValue)
+  }
+  
+  const handleCopyField = async (e, field) => {
+    e.preventDefault()
+    if (!importedData) return
+
+    if (field === 'biography') {
+      biographyRef?.current?.setEditorValue(importedData.biography.markdownValue)
+      return
+    } else if (field === 'howOthersCanHelpMe') {
+      howOthersCanHelpMeRef?.current?.setEditorValue(importedData.howOthersCanHelpMe.markdownValue)
+      return
+    } else if (field === 'howICanHelpOthers') {
+      howICanHelpOthersRef?.current?.setEditorValue(importedData.howICanHelpOthers.markdownValue)
+      return
+    // } else if (field === 'organizerOfGroupIds') {
+    //   // TODO: this field needs more work -
+    //   // prob need to only list groups they are assigned to on the forum, and link to the form for requests
+    //   setFormValues({
+    //     ...formValues,
+    //     [field]: importedData.organizerOfGroups.map(g => g._id)
+    //   })
+    //   return
+    } else if (field === 'mapLocation') {
+      const location = await importMapLocation()
+      if (location) {
+        setFormValues({
```

Comment:
JavaScript closures are hard to reason about, and I think referencing a global variable in this way can lead to race conditions. I think instead this should be
```
setFormValues(currentValues => {
      ...currentValues ,
      mapLocation: location
})
```
same comment about all the other references to useState setters